include/s3select_functions.h

```diff
@@ -1188,70 +1189,232 @@ struct _fn_in : public base_function
-    transform(string_value);
-    compiled_regex = std::regex(string_value);
-  } 
-
-  void transform(std::string& s)
+  const char *p = s.c_str();
+  std::vector<char> v;
+  
+  while(*p)
   {
-    std::string::size_type i = 0;
-    while (!s.empty())
+    switch (st)
     {
-        i = s.find("%%", i);
-        if (i == std::string::npos) break;
-        s.erase(i, 1);      
-    }
-    bool startswith = (s[0] == '%' && s[s.size()-1] != '%');
-    bool endswith = (s[0] != '%' && s[s.size()-1] == '%');
-    bool startend = (s[0] == '%' && s[s.size()-1] == '%');
-    if (startswith) {
-      std::replace(s.begin(), s.begin()+1, '%', '^');
-      s.insert(1, ".");
-      s.insert(2, "*");
-      s.append("$");
-      std::replace(s.begin(), s.end(), '_', '.');
-    } else if (endswith) {
-      s.insert(0, "^");
-      std::replace(s.end()-1, s.end(), '%', '$');
-      s.insert(s.size()-1,".");
-      s.insert(s.size()-1,"*");
-      std::replace(s.begin(), s.end(), '_', '.');
-    } else if (startend) {
-      std::replace(s.begin(), s.begin()+1, '%', '^');
-      s.insert(1, ".");
-      s.insert(2, "*");
-      std::replace(s.end()-1, s.end(), '%', '$');
-      s.insert(s.size()-1,".");
-      s.insert(s.size()-1,"*");
-      std::replace(s.begin(), s.end(), '_', '.');
-    } else {
-      std::string findThis = "%";
-      std::string replaceWith = ".*";
-      s.insert(0,"^");
-      s.append("$");
-      std::size_t pos = 0;
-      while ((pos = s.find(findThis, pos)) != std::string::npos)
-      {
-        s.replace(pos, findThis.length(), replaceWith);
-        pos += replaceWith.length();
-      }
-      std::replace(s.begin(), s.end(), '_', '.');
+      case START:
+          if(*p == escape) {
+          st = ESCAPE;
+          v.push_back('^');
+          } else if(*p == '%') {
+        v.push_back('^');
+            v.push_back('.');
+            v.push_back('*');
+            st = START_STAR_CHAR;
+            } else if(*p == '_') {
+        v.push_back('^');
+        v.push_back('.');
+        st=START_METACHAR;
+          } else {
+        v.push_back('^');
+        v.push_back(*p);
+        st=START_ANYCHAR;
+          } break;
+
+      case START_STAR_CHAR:
+          if(*p == escape) {
+          st = ESCAPE;
+          } else if(*p == '%') {
+        st = START_STAR_CHAR;
+        } else if(*p == '_') {
+        v.push_back('.');
+        st = METACHAR;
+        } else {
+        v.push_back(*p);
+        st = ANYCHAR;
+        } break;
+
+      case START_METACHAR:
+          if(*p == escape) {
+          st = ESCAPE;
+          } else if(*p == '_') {
+        v.push_back('.');
+        st = METACHAR;
+        } else if(*p == '%') {
+        v.push_back('.');
+        v.push_back('*');
+        st = STAR_CHAR;
+        } else {
+        v.push_back(*p);
+        st = ANYCHAR;
+        } break;
+
+      case START_ANYCHAR:
+          if(*p == escape) {
+          st = ESCAPE;
+          } else if(*p == '_' && !(*(p + 1))) {
```

Comment:
it's better to replace /!(*(p + 1))/  with boolean last_char, it's more readable.